src/mbgl/style/expression/in.cpp

```diff
@@ -0,0 +1,113 @@
+#include <mbgl/style/expression/type.hpp>
+#include <mbgl/util/string.hpp>
+
+namespace mbgl {
+namespace style {
+namespace expression {
+
+namespace {
+bool isComparableType(type::Type type) {
+    return type == type::Boolean || type == type::String || type == type::Number || type == type::Null ||
+           type == type::Value;
+}
+
+bool isComparableRuntimeType(type::Type type) {
+    return type == type::Boolean || type == type::String || type == type::Number || type == type::Null;
+}
+
+bool isSearchableRuntimeType(type::Type type) {
+    return type == type::String || type.is<type::Array>() || type == type::Null;
+}
+} // namespace
+
+In::In(std::unique_ptr<Expression> needle_, std::unique_ptr<Expression> haystack_)
+    : Expression(Kind::In, type::Boolean), needle(std::move(needle_)), haystack(std::move(haystack_)) {
+    assert(isComparableType(needle->getType()));
+    assert(isSearchableRuntimeType(haystack->getType()) || haystack->getType() == type::Value);
+}
+
+EvaluationResult In::evaluate(const EvaluationContext& params) const {
+    const EvaluationResult evaluatedHaystack = haystack->evaluate(params);
+    if (!evaluatedHaystack) {
+        return evaluatedHaystack.error();
+    }
+
+    const EvaluationResult evaluatedNeedle = needle->evaluate(params);
+    if (!evaluatedNeedle) {
+        return evaluatedNeedle.error();
+    }
+
+    type::Type evaluatedNeedleType = typeOf(*evaluatedNeedle);
+    if (!isComparableRuntimeType(evaluatedNeedleType)) {
+        return EvaluationError{"Expected first argument to be of type boolean, string or number, but found " +
+                               toString(evaluatedNeedleType) + " instead."};
+    }
+
+    type::Type evaluatedHaystackType = typeOf(*evaluatedHaystack);
+    if (!isSearchableRuntimeType(evaluatedHaystackType)) {
+        return EvaluationError{"Expected second argument to be of type array or string, but found " +
+                               toString(evaluatedHaystackType) + " instead."};
+    }
+
+    if (evaluatedNeedleType == type::Null || evaluatedHaystackType == type::Null) {
+        return EvaluationResult(false);
+    }
+
+    if (evaluatedHaystackType == type::String) {
+        const auto haystackString = evaluatedHaystack->get<std::string>();
+        std::string needleValue = "";
+        if (evaluatedNeedleType == type::Boolean) {
+            needleValue = evaluatedNeedle->get<bool>() ? "true" : " false";
+        } else if (evaluatedNeedleType == type::String) {
+            needleValue = evaluatedNeedle->get<std::string>();
+        } else if (evaluatedNeedleType == type::Number) {
+            needleValue = std::to_string(evaluatedNeedle->get<double>());
+            needleValue.erase(needleValue.find_last_not_of('0') + 1, std::string::npos);
+            needleValue.erase(needleValue.find_last_not_of('.') + 1, std::string::npos);
+        }
+        return EvaluationResult(haystackString.find(needleValue) != std::string::npos);
```

Comment:
I think this whole block can be rewritten as:

``` c++
const auto haystackString = evaluatedHaystack->get<std::string>();
const auto needleString = toString(evaluatedNeedle);
return EvaluationResult(haystackString.find(needleString) != std::string::npos);
```

Reply:
Awesome!